### PR TITLE
[ONNX][re-open] onnx interpreter 

### PR DESCRIPTION
**Pull request by baek2sm**
I accidentally closed the original PR #3338 (it was automatically closed after a force‑push with no diff, and it can no longer be reopened. it was my mistake, I'm sorry for reviewer @myungjoo and @songgot ), so I opened a new PR. I've fixed some issues that were causing the unit tests to fail.

This PR contains the commits that handle the ONNX model compilation process. The work from SRI‑B built on top of these commits results in PR #3571.
- This PR: for compilation
- PR 3571: for prefill inference

This PR contains multiple commits that get the LLM model to a compile‑ready state, including:
- Adding support for loading properties from ONNX operators.
- Adding support for loading properties from ONNX tensors.
- Implementing required layers such as Cast, Slice, and others.
- Adding documentation for ONNX.

The execution result of the example app is as follows (compiled llama 1b, without inference):
```
================================================================================
          Layer name          Layer type    Output dimension         Input layer
================================================================================
         onnx__add_6               input             1:1:1:1
--------------------------------------------------------------------------------
 onnx__add_6/generat            multiout             1:1:1:1         onnx__add_6
--------------------------------------------------------------------------------
                 sin               input            1:1:1:64
--------------------------------------------------------------------------------
 sin/generated_out_0            multiout            1:1:1:64                 sin
--------------------------------------------------------------------------------

...

--------------------------------------------------------------------------------
   model_norm_cast_1                cast          1:1:1:2048      model_norm_mul
--------------------------------------------------------------------------------
    model_norm_mul_1            multiply          1:1:1:2048   model_norm_weight
                                                              model_norm_cast_1
--------------------------------------------------------------------------------
      lm_head_matmul              matmul         1:1:1:50304    model_norm_mul_1
                                                               onnx__matmul_3531
================================================================================
```

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

Signed-off-by: Seungbaek Hong <sb92.hong@samsung.com>

---

> [!NOTE]
> Broaden ONNX op support and graph deserialization, add Cast/Gather/Slice/Negative layers, simplify/refactor Weight layer, and update docs, examples, enums, and tests.
> 
> - **ONNX Interpreter**:
>   - Implement graph deserialization with input/weight loading (dtype/name), constant handling, and output remapping.
>   - Add handlers for `Add/Sub/Mul/Div/MatMul`, unary ops (`Pow/Sqrt/Softmax/Sigmoid/Relu/Identity/Cosine/Sine/Tangent/Neg`), and new ops: `Reshape`, `Transpose`, `Gather`, `Cast`, `Slice`, `Concat`, `ReduceMean`.
>   - Introduce attribute/tensor-attribute extraction helpers and ONNX→NNTrainer dtype mapping.
> - **New/Updated Layers**:
>   - Add `cast`, `gather`, `slice`, and `negative` layers with forward/backward implementations and registration in `AppContext`.
>   - Replace `neg` with `negative` (API helpers and tests updated).
>   - Refactor `weight` layer to single tensor API; add props `tensor_dtype` and `weight_name`; disable backwarding.
> - **API/Enums**:
>   - Extend `ml_train_layer_type_e` and C++ `LayerType` with `CAST/GATHER/SLICE/NEG` and adjust helpers.
>   - Add common properties `StartIndex` and `EndIndex`; exporter awareness updated.
> - **Docs & Example**:
>   - Add `docs/how-to-run-onnx-model.md` with build/run instructions.
>   - Example app now loads `llama_example.onnx`.
> - **Tests/Build**:
>   - Add unit tests for `cast` and `negative`; update model tests to use `negative`.
>   - Update meson build lists for new layers and tests.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 5044c2e39df3ee945467f3be1338d72b7708c5d5. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: api/nntrainer-api-common.h [hunk 1]
```diff
@@ -72,17 +72,24 @@ typedef enum {
...
+  ML_TRAIN_LAYER_TYPE_REDUCE_SUM = 49, /**< ReduceSum Layer type (Since 9.0) */
   ML_TRAIN_LAYER_TYPE_CHANNEL_SHUFFLE =
-    45,                         /**< Channel Shuffle Layer type (Since 9.0)*/
-  ML_TRAIN_LAYER_TYPE_NEG = 46, /**< Neg Layer type (Since 9.0)*/
-  ML_TRAIN_LAYER_TYPE_REDUCE_SUM = 48, /**< ReduceSum Layer type (Since 9.0) */
+    50, /**< Channel Shuffle Layer type (Since 9.0)*/
+  ML_TRAIN_LAYER_TYPE_REDUCE_MEAN =
+    51,                                /**< ReduceMean Layer type (Since 9.0) */
+  ML_TRAIN_LAYER_TYPE_RESHAPE = 52,    /**< Reshape Layer type (Since 9.0) */
+  ML_TRAIN_LAYER_TYPE_UPSAMPLE2D = 53, /**< Upsample2D Layer type (Since 9.0) */
```
Comment (myungjoo) [inline api/nntrainer-api-common.h]: Keep the API forward compatibility.
1. Do not rename NEG to NEGATIVE. This breaks compiling of older applications
2. Do not change number for pre-existing enums. This breaks older binaries.

File: api/ccapi/include/layer.h [hunk 2]
```diff
@@ -111,19 +114,23 @@ enum LayerType {
...
-    ML_TRAIN_LAYER_TYPE_CHANNEL_SHUFFLE,      /**< Channel Shuffle Layer type */
+  LAYER_REDUCE_MEAN =
+    ML_TRAIN_LAYER_TYPE_REDUCE_MEAN,           /**< Reduce mean Layer type */
+  LAYER_RESHAPE = ML_TRAIN_LAYER_TYPE_RESHAPE, /**< Reshape Layer type */
+  LAYER_UPSAMPLE2D =
+    ML_TRAIN_LAYER_TYPE_UPSAMPLE2D, /**< Upsample 2D Layer type */
+
+  LAYER_TIME_DIST = 1000,                /**< Time Distributed Layer type */
+  LAYER_BACKBONE_TFLITE = 1001,          /**< Backbone using TFLite */
+  LAYER_LOSS_CONSTANT_DERIVATIVE = 1002, /**< Synthetic loss layer to feed
```
Comment (myungjoo) [inline api/ccapi/include/layer.h]: 1. If the values are correct (not duplicated), do not change the enum values. It breaks applications.
2. You may update the values if they have been incorrect (duplicated); however, this should be clearly states at the release note of next version; you are breaking APIs. You will need to update the nntrainer version with this. You do not need to re-release with next version after this PR; but you need to remember that this is required and update release notes in the next version release.

Comment (baek2sm) [reply api/ccapi/include/layer.h]: These layers are not released as APIs

File: api/nntrainer-api-common.h (hunk 1 above)
Comment (baek2sm) [reply api/nntrainer-api-common.h]: I only edited the layers that haven’t been merged into the release branch yet.
However, I’ll make the changes according to your feedback.

File: nntrainer/compiler/onnx_interpreter.cpp [hunk 3]
```diff
@@ -12,35 +12,233 @@
...
+    handleUnaryOp(node, rep, layerKeyMap[node.op_type()], props);
+  };
+
+  NodeHandlers["Slice"] = [this](const onnx::NodeProto &node,
+                                 GraphRepresentation &rep) {
+    std::vector<std::string> props;
+    const std::string &data = node.input(0);
+    const std::string &start = node.input(1);
+    const std::string &end = node.input(2);
+    const std::string &axis = node.input(3);
```
Comment (myungjoo) [inline nntrainer/compiler/onnx_interpreter.cpp:198]: Please check if these node names are ok, especially with slice/gather. Check if node.output(n) is appropriate instead.

File: nntrainer/layers/gather_layer.cpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+      "The axis property of GatherLayer should be between 1 and 3.");
+  }
+
+  if (inputDim[0] != indexDim[0]) {
+    throw std::invalid_argument(
+      "The batch size of the input and index should be same.");
+  }
+
+  TensorDim outputDim =
+    TensorDim(indexDim.batch(), indexDim.width(), inputDim.width());
```
Comment (myungjoo) [inline nntrainer/layers/gather_layer.cpp]: Is this really ok? ( https://onnx.ai/onnx/operators/onnx__Gather.htm )

File: nntrainer/compiler/onnx_interpreter.cpp [hunk 5]
```diff
@@ -12,35 +12,233 @@
...
+                                     GraphRepresentation &rep) {
+    std::vector<std::string> props;
+    props.push_back(extractAttribute(node, "perm", "direction", ",", 1));
+    handleUnaryOp(node, rep, layerKeyMap[node.op_type()], props);
+  };
+
+  NodeHandlers["Gather"] = [this](const onnx::NodeProto &node,
+                                  GraphRepresentation &rep) {
+    std::vector<std::string> props;
+    props.push_back("axis=3");
```
Comment (myungjoo) [inline nntrainer/compiler/onnx_interpreter.cpp]: uh.. can you statically assume "axis=3"?

File: nntrainer/compiler/onnx_interpreter.cpp [hunk 6]
```diff
@@ -12,35 +12,233 @@
...
+      extractTensorAttribute(constantTensors.at(axis), "value", "axis"));
+
+    handleUnaryOp(node, rep, layerKeyMap[node.op_type()], props);
+  };
+
+  NodeHandlers["Concat"] = [this](const onnx::NodeProto &node,
+                                  GraphRepresentation &rep) {
+    std::vector<std::string> props;
+    props.push_back(extractAttribute(node, "axis", "axis"));
+    handleBinaryOp(node, rep, layerKeyMap[node.op_type()], props);
```
Comment (myungjoo) [inline nntrainer/compiler/onnx_interpreter.cpp:220]: Be careful.

ONNX Concat accepts N tensors, not 2 tensors. If you apply handleBinaryOp here, you need an error handler that prepares for N>2 input tensors.

File: docs/how-to-run-onnx-model.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+#include <model.h>
+#include <nntrainer-api-common.h>
+#include <optimizer.h>
+#include <util_func.h>
+
+int main() {
+  auto model = ml::train::createModel();
+
+  try {
+    std::string path = "../../../../Applications/ONNX/jni/add_example.onnx";
```
Comment (myungjoo) [inline docs/how-to-run-onnx-model.md]: You don't have add_example.onnx anymore.

File: nntrainer/compiler/onnx_interpreter.cpp (hunk 3 above)
Comment (baek2sm) [reply nntrainer/compiler/onnx_interpreter.cpp:198]: Good point. If an optional input is left empty, it can cause problems. I modified it so that it will raise an error whenever any input is missing(When the model is exported from PyTorch, all inputs are always present.)

File: nntrainer/compiler/onnx_interpreter.cpp (hunk 6 above)
Comment (baek2sm) [reply nntrainer/compiler/onnx_interpreter.cpp:220]: It is currently implemented to work only when there are two inputs, so I added error handling for case where the input length is not 2. I'll update to support multiple inputs in a later commit.

File: nntrainer/compiler/onnx_interpreter.cpp (hunk 5 above)
Comment (baek2sm) [reply nntrainer/compiler/onnx_interpreter.cpp]: It was a test code. thank you a lot! I modified it.

File: docs/how-to-run-onnx-model.md (hunk 7 above)
Comment (baek2sm) [reply docs/how-to-run-onnx-model.md]: Thanks, I updated the document based on the llama example.

File: nntrainer/layers/gather_layer.cpp (hunk 4 above)
Comment (baek2sm) [reply nntrainer/layers/gather_layer.cpp]: I'll take a closer look at this part on Monday.

File: api/ccapi/include/layer.h [hunk 8]
```diff
@@ -111,19 +111,23 @@ enum LayerType {
...
-  LAYER_TRANSPOSE = ML_TRAIN_LAYER_TYPE_TRANSPOSE, /**< Transpose Layer type */
-  LAYER_CHANNEL_SHUFFLE =
-    ML_TRAIN_LAYER_TYPE_CHANNEL_SHUFFLE,      /**< Channel Shuffle Layer type */
+  LAYER_REDUCE_MEAN =
+    ML_TRAIN_LAYER_TYPE_REDUCE_MEAN,           /**< Reduce mean Layer type */
+  LAYER_RESHAPE = ML_TRAIN_LAYER_TYPE_RESHAPE, /**< Reshape Layer type */
+  LAYER_UPSAMPLE2D =
+    ML_TRAIN_LAYER_TYPE_UPSAMPLE2D, /**< Upsample 2D Layer type */
+
+  LAYER_TIME_DIST = 1000,                /**< Time Distributed Layer type */
```
Comment (songgot) [inline api/ccapi/include/layer.h]: If '1000' doesn't mean much, wouldn't it be okay to start the value at 900 since the 'LAYER_UNKNOWN' value is 999?

File: nntrainer/layers/gather_layer.cpp (hunk 4 above)
Comment (baek2sm) [reply nntrainer/layers/gather_layer.cpp]: fixed it! thanks

**Comment by myungjoo** [comment]
@cursor Review this pull request. Check for logical / performance bugs. For each issue, make a suggestion of code changes in the format of code diff.

File: nntrainer/layers/gather_layer.cpp [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+namespace nntrainer {
+
+void GatherLayer::finalize(InitLayerContext &context) {
+  axis = std::get<props::Axis>(gather_props).get();
+  TensorDim inputDim = context.getInputDimensions()[0];
+  TensorDim indexDim = context.getInputDimensions()[1];
+
+  if (axis < 1 || axis > 3) {
+    throw std::invalid_argument(
+      "The axis property of GatherLayer should be between 1 and 3.");
```
Comment (myungjoo) [inline nntrainer/layers/gather_layer.cpp:32]: Is this really a valid condition? no axis=0 option?

Comment (baek2sm) [reply nntrainer/layers/gather_layer.cpp:32]: We can also support `axis=0`, but for us, `axis=0` always means batch dimension, and changing the size of that batch dimension dynamically can lead to some unintentional issues(it can cause mismatching batch sizes between inputs and labels.. so we need to consider more cases if we want to support gathering operation for batch dimension).

File: nntrainer/compiler/onnx_interpreter.cpp [hunk 10]
```diff
@@ -12,35 +12,249 @@
...
+                                    std::vector<std::string> &props) {
+  std::vector<std::string> inputNames = createOutputRemap(node);
+  props.push_back("name=" + cleanName(node.name()));
+  if (op_type == "activation") {
+    props.push_back("activation=" + activationKeyMap[node.op_type()]);
+  }
+  props.push_back("input_layers=" + inputNames[0]);
+
+  representation.push_back(
+    createLayerNode(op_type, {props.begin(), props.end()}));
```
Comment (jijoongmoon) [inline nntrainer/compiler/onnx_interpreter.cpp:103]: It is OK for now. (Not this PR). But I think we don't need to create actual nntrainer layer node for now. Because we also need to consider the delegate, the created layer could be meanless at this converting pass.

File: nntrainer/compiler/onnx_interpreter.h [hunk 11]
```diff
@@ -36,14 +40,58 @@ class ONNXInterpreter : public GraphInterpreter {
    * @brief Construct a new ONNXInterpreter object
    *
    */
-  ONNXInterpreter(){};
+  ONNXInterpreter() { registerNodeHandlers(); };
```
Comment (jijoongmoon) [inline nntrainer/compiler/onnx_interpreter.h:44]: I think the ONNXInterpreter should have only one registry to handle these NodeHandler. May be you could use singleton.

**Review by djeong20 — APPROVED** [review]

**Review by haehun — APPROVED** [review]
It would be pretty good to use it to convert models to NPU.

**Comment by baek2sm** [comment]
I’ll force‑push the rebased changes tomorrow because the network(proxy) issue for now.
